neuro_data/movies/stats.py

```diff
@@ -146,3 +149,276 @@ def _make_tuples(self, key):
+        for i, index in enumerate(target_indices):
+            inputs[i] = dataset[index].inputs
+        assert np.all(np.abs(np.diff(inputs, axis=0)) == 0), 'Images of oracle trials do not match'
+
+    def sample_frames_from_dataset(self, index, dataset, min_frames, num_of_neurons):
+        # Compute start_index
+        starting_index = 0
+        upper_bound = dataset[index].responses.shape[0] - min_frames
+    
+        if upper_bound != 0:
+            starting_index = np.random.randint(0, upper_bound)
+            
+        return dataset[index].responses[starting_index:starting_index+min_frames]
+
+    def compute_oracle(self, outputs):
+        r = outputs.shape[0]
+        mu = outputs.mean(axis=0, keepdims=True)
+        oracles = ((mu - outputs / r) * r / (r - 1)).reshape(-1, outputs.shape[-1])
+        return oracles
+
+    def sample_and_compute_oracle(self, dataset, condition_hashes, sample_size, min_frames):
+        num_of_neurons = dataset[0].responses.shape[1]
+        
+        # Oracle compuatation
+        true_responses = np.empty(shape=[len(condition_hashes), sample_size * min_frames, num_of_neurons])
+        true_oracles = np.empty(shape=[len(condition_hashes), sample_size * min_frames, num_of_neurons])
+        
+        null_responses = np.empty(shape=[len(condition_hashes), sample_size * min_frames, num_of_neurons])
+        null_oracles = np.empty(shape=[len(condition_hashes), sample_size * min_frames, num_of_neurons])
+        
+        response_matrix = np.empty(shape=[sample_size, min_frames, num_of_neurons]) # this is a temp storage for data extraction
+        
+        for i in range(0, len(condition_hashes)):
+            num_of_neurons = dataset[0].responses.shape[1]
+            dataset_condition_hashes = dataset.condition_hashes
+            # Oracle compuatation
+            true_responses = np.empty(shape=[len(condition_hashes), sample_size * min_frames, num_of_neurons])
+            true_oracles = np.empty(shape=[len(condition_hashes), sample_size *  min_frames, num_of_neurons])
+            
+            null_responses = np.empty(shape=[len(condition_hashes), sample_size * min_frames, num_of_neurons])
+            null_oracles = np.empty(shape=[len(condition_hashes), sample_size *  min_frames, num_of_neurons])
+            
+            response_matrix = np.empty(shape=[sample_size, min_frames, num_of_neurons]) # this is a temp storage for data extraction
+            
+            for i in range(0, len(condition_hashes)):
+                # True Oracle Computation
+                # For each condition_hashes, sample (sample_size) trials to construct the true_response_matrix
+                # Select (sample_size) trials
+                
+                true_target_indices = self.sample_from_condition_hash(condition_hashes[i], dataset_condition_hashes, sample_size)
+                self.check_input(true_target_indices, dataset)
```

Comment:
this should be after you subsample the frames

Reply:
also, the check_input function won't work because you're no longer indexing and evaluating the entire response arrays, just the frames of interest